libimage/pull.go

```diff
@@ -25,36 +26,70 @@ import (
+	// scattering the logic of short-name modes; it should remain in
+	// c/image.
+	//
+	// Hence, this function will return all possible candidates and
+	// podman-remote will implicitly make the right decision by calling
+	// shortnames.Prompt().
 
-	// If true, all tags of the image will be pulled from the container
-	// registry.  Only supported for the docker transport.
-	AllTags bool
-}
+	mode, err := sysregistriesv2.GetShortNameMode(&r.systemContext)
+	if err != nil {
+		return nil, -1, err
+	}
+	possiblyUnqualifiedName, ref, localImage, err := r.parsePossiblyUnqualifiedName(name, pullPolicy)
+	if err != nil {
+		return nil, -1, err
+	}
 
-// Pull pulls the specified name.  Name may refer to any of the supported
-// transports from github.com/containers/image.  If no transport is encoded,
-// name will be treated as a reference to a registry (i.e., docker transport).
-//
-// Note that pullPolicy is only used when pulling from a container registry but
-// it *must* be different than the default value `config.PullPolicyUnsupported`.  This
-// way, callers are forced to decide on the pull behaviour.  The reasoning
-// behind is that some (commands of some) tools have different default pull
-// policies (e.g., buildah-bud versus podman-build).  Making the pull-policy
-// choice explicit is an attempt to prevent silent regressions.
-//
-// The error is storage.ErrImageUnknown iff the pull policy is set to "never"
-// and no local image has been found.  This allows for an easier integration
-// into some users of this package (e.g., Buildah).
-func (r *Runtime) Pull(ctx context.Context, name string, pullPolicy config.PullPolicy, options *PullOptions) ([]*Image, error) {
-	logrus.Debugf("Pulling image %s (policy: %s)", name, pullPolicy)
+	// Already resolved to a local image.  Only possible when it clearly
+	// refers to a local one by digest.
+	if localImage != nil {
+		return []string{possiblyUnqualifiedName}, mode, nil
+	}
 
-	if options == nil {
-		options = &PullOptions{}
+	switch ref.Transport().Name() {
+	case registryTransport.Transport.Name():
+		// All good
+	default:
+		return nil, -1, fmt.Errorf("unsupported transport for image resolution: %q (does not refer to a registry)", ref.Transport().Name())
 	}
 
+	shortNameModeDisabled := types.ShortNameModeDisabled // we want to either fetch the alias or all search registries
+	options := &PullOptions{
+		copyFromRegistryDryRun: true,
+		shortNameMode:          &shortNameModeDisabled,
+	}
+
+	candidates, err := r.copyFromRegistry(ctx, ref, possiblyUnqualifiedName, pullPolicy, options)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	return candidates, mode, nil
+}
+
+// parsePossiblyUnqualifiedName parses the specified name into four components:
+// 1) A (possibly unqualified) image name
+// 2) A reference in possibly any supported transport or nil
+// 3) An image when the name cleary refers to a local one (*must* be used if set)
+// 4) An error
```

Comment:
Is this correct?
> If the input refers to a registry, the returned reference is a registryTransport reference, but the caller should ignore it (apart from checking for registryTransport) and use the “(possibly unqualified) image name” return value instead.

(Maybe the API should be restructured instead of documented, but, meh, it’s file-private with two users, let’s not worry about that now with all the short name work.)